### ISSUE TITLE
Using candidate sampling probability to explicitly set negative examples in Retrieval task

### ISSUE THREAD

**Issue by nialloh23** [tensorflow/recommenders#232]
**Context:**  
I am training a two tower candidate retrieval model on examples of user_sku pairs which fit the user (the skus are fashion styles of a certain size). The ultimate goal is to establish a set of user + sku embeddings which represent the users and skus sizes.  

For this example I don't believe it makes sense to sample negative examples from all of the other skus (e.g. a sku size 8 dress is not really comparable to a size 3 shoe). In reality I want to use the non-purchased skus of the same product as explicit negative examples. **Example: for a product ABC that fit in size 2 (ABC_2), I want to use the other sizes for this product as explicit negative examples (ABC_0, ABC_4, ABC_6, ABC8)** . I've tried to sketch out the example below.

![Screenshot 2021-02-18 at 12 01 51](https://user-images.githubusercontent.com/9644540/108354293-28bd3180-71e1-11eb-8dbb-72278e4f792d.png)

**Problem/Question:**  
- I was wondering if it is possible to pass in a list of candidate_ids to the retrieval task to use as explicit negatives to train with (e.g. **fit**: ABC_2, **didnt_fit:**[ABC_0, ABC_4, ABC_6, ABC8])?   

- I was thinking that there may be a way to use the `candidate_sampling_probability` to explicitly state for each training example that the sampling probability of the explicit negative examples should be 1 and all other candidates should be set to 0? does this make sense?.....if there are 100,000 candidate skus does my candidate sampling probability tensor need to include probabilities for all 100k candidates e.g. 4 candidates =1, the reset set to 0) and can these probabilities be different for every positive example? (hope this makes sense!). 

- I had initially thought I could use the `candidate_ids` parameter to explicitly set these negative examples but when I looked at the source code it looked like these candidate_ids are used to remove ids of candidates that clash with the positive example. I'm not sure if I'm interpreting this correctly or if there is indeed a way to use this candidate_ids input as a way to explicitly set the neg. candidates?

If helpful for additional context this is the paper I'm reproducing: http://ceur-ws.org/Vol-2410/paper13.pdf

**Comment by maciejkula** [comment]
Are you trying to model size/colour preferences? The resulting model will discriminate well between variants of a single SKU, but may be very poor at modelling between-SKU preferences. That may or may not work for you, depending on your objective.

Either way, unfortunately we do not support something like this out of the box right now - making negative selection a lot more flexible is on our roadmap, but probably isn't going to come fast enough for you.

It shouldn't be too difficult to put together out of basic TF components. If you have some implementation sketches I'd be happy to give some feedback in this thread.

**Comment by nialloh23** [comment]
Hi @maciejkula. Thanks for your reply.
Correct, I'm trying to model size preference. The goal is to suggest the correct sku size (child) within a given style(parent).
Example: for style ABC with sku sizes ABC_2, ABC_4, ABC_6, ABC_8-> I want to to reccomend the sku size that will fit a user best.

My positive interactions in my candidate model are user:sku pairs that fit users based on feedback.
- My queries are: user_ids
- My candidates are: sku_ids 
I'm using the candidate model to successfully create user & sku embeddings which are representative of their true size.
I then take the user & sku embeddings and get their cosine similarity scores. The user_sku pair with the lowest cosine similarity score is recommended to the user.

I'm getting quite good results already (67% accuracy on predicting sku sizes that fit). Typical industry benchmarks indicate results in the range 60-75% are typical for this task.

**I have one question:**
I have eyeballed the mispredicted size recommendations and my wrong predictions all seem to bias in one direction (E.g. more often than not my wrong predictions are recommending a smaller size than I should have). I've done some research on this topic and I believe it may have something to do with the negative sampling strategy adopted. Is it correct to say that the default in-batch softmax approach is biased towards the positive distribution?  I think this may be the source of my errors.

I see in a previous issue you mentioned that a 'sampled softmax' strategy samples uniformly from all candidates. Is it possible to adopt this approach in TFRS yet? 

Thanks as again for your work on this and your patience in responding to early issues!
